File: android/app/src/main/kotlin/com/nguyenduc/edtech/ed_tech/EduTechNativeAdFactory.kt
package com.nguyenduc.edtech.ed_tech

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.TextView
import com.google.android.gms.ads.nativead.MediaView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin

class EduTechNativeAdFactory : GoogleMobileAdsPlugin.NativeAdFactory {

    override fun createNativeAd(
        nativeAd: NativeAd,
        customOptions: MutableMap<String, Any>?
    ): NativeAdView {
        val activity = MainActivity.currentActivity ?: throw IllegalStateException("Activity not available")
        val inflater = LayoutInflater.from(activity)
        val nativeAdView = inflater.inflate(R.layout.native_ad_layout, null) as NativeAdView

        // Attribution
        nativeAdView.findViewById<TextView>(R.id.native_ad_attribution).text = "Ad"

        // Headline
        val headlineView = nativeAdView.findViewById<TextView>(R.id.native_ad_headline)
        headlineView.text = nativeAd.headline
        nativeAdView.headlineView = headlineView

        // Body
        val bodyView = nativeAdView.findViewById<TextView>(R.id.native_ad_body)
        if (nativeAd.body != null) {
            bodyView.text = nativeAd.body
            bodyView.visibility = View.VISIBLE
            nativeAdView.bodyView = bodyView
        } else {
            bodyView.visibility = View.GONE
        }

        // Icon
        val iconView = nativeAdView.findViewById<ImageView>(R.id.native_ad_icon)
        if (nativeAd.icon != null) {
            iconView.setImageDrawable(nativeAd.icon!!.drawable)
            nativeAdView.iconView = iconView
        } else {
            iconView.visibility = View.GONE
        }

        // Media
        val mediaView = nativeAdView.findViewById<MediaView>(R.id.native_ad_media)
        nativeAdView.mediaView = mediaView
        mediaView.setMediaContent(nativeAd.mediaContent)

        // Advertiser
        val advertiserView = nativeAdView.findViewById<TextView>(R.id.native_ad_advertiser)
        if (nativeAd.advertiser != null) {
            advertiserView.text = nativeAd.advertiser
            advertiserView.visibility = View.VISIBLE
            nativeAdView.advertiserView = advertiserView
        } else {
            advertiserView.visibility = View.GONE
        }

        // CTA
        val ctaView = nativeAdView.findViewById<Button>(R.id.native_ad_cta)
        if (nativeAd.callToAction != null) {
            ctaView.text = nativeAd.callToAction
            nativeAdView.callToActionView = ctaView
        } else {
            ctaView.visibility = View.GONE
        }

        // Store
        val storeView = nativeAdView.findViewById<TextView>(R.id.native_ad_store)
        if (nativeAd.store != null) {
            storeView.text = nativeAd.store
            storeView.visibility = View.VISIBLE
            nativeAdView.storeView = storeView
        } else {
            storeView.visibility = View.GONE
        }

        // Price
        val priceView = nativeAdView.findViewById<TextView>(R.id.native_ad_price)
        if (nativeAd.price != null) {
            priceView.text = nativeAd.price
            priceView.visibility = View.VISIBLE
            nativeAdView.priceView = priceView
        } else {
            priceView.visibility = View.GONE
        }

        // Star Rating
        val starRatingView = nativeAdView.findViewById<RatingBar>(R.id.native_ad_stars)
        if (nativeAd.starRating != null) {
            starRatingView.rating = nativeAd.starRating!!.toFloat()
            starRatingView.visibility = View.VISIBLE
            nativeAdView.starRatingView = starRatingView
        } else {
            starRatingView.visibility = View.GONE
        }

        nativeAdView.setNativeAd(nativeAd)
        return nativeAdView
    }
}

File: ios/Runner/EduTechNativeAdFactory.swift
import GoogleMobileAds
import UIKit

/// Native Ad Factory for Edu-Tech iOS.
///
/// Builds the native ad view programmatically — no XIB dependency.
class EduTechNativeAdFactory: FLTNativeAdFactory {

    func createNativeAd(
        _ nativeAd: GADNativeAd,
        customOptions: [AnyHashable: Any]? = nil
    ) -> GADNativeAdView? {
        let nativeAdView = GADNativeAdView()
        nativeAdView.translatesAutoresizingMaskIntoConstraints = false
        nativeAdView.backgroundColor = .white
        nativeAdView.layer.cornerRadius = 8
        nativeAdView.layer.borderWidth = 1
        nativeAdView.layer.borderColor = UIColor(white: 0.88, alpha: 1).cgColor
        nativeAdView.clipsToBounds = true

        // ----- Attribution -----
        let attributionLabel = UILabel()
        attributionLabel.text = "Ad"
        attributionLabel.font = UIFont.boldSystemFont(ofSize: 10)
        attributionLabel.textColor = UIColor(red: 1, green: 0.6, blue: 0, alpha: 1)
        attributionLabel.translatesAutoresizingMaskIntoConstraints = false

        // ----- Icon -----
        let iconView = UIImageView()
        iconView.contentMode = .scaleAspectFit
        iconView.translatesAutoresizingMaskIntoConstraints = false
        if let icon = nativeAd.icon {
            iconView.image = icon.image
        }
        nativeAdView.iconView = iconView

        // ----- Headline -----
        let headlineLabel = UILabel()
        headlineLabel.font = UIFont.boldSystemFont(ofSize: 16)
        headlineLabel.textColor = UIColor(white: 0.13, alpha: 1)
        headlineLabel.numberOfLines = 2
        headlineLabel.translatesAutoresizingMaskIntoConstraints = false
        headlineLabel.text = nativeAd.headline
        nativeAdView.headlineView = headlineLabel

        // ----- Advertiser -----
        let advertiserLabel = UILabel()
        advertiserLabel.font = UIFont.systemFont(ofSize: 12)
        advertiserLabel.textColor = UIColor(white: 0.46, alpha: 1)
        advertiserLabel.numberOfLines = 1
        advertiserLabel.translatesAutoresizingMaskIntoConstraints = false
        if let advertiser = nativeAd.advertiser {
            advertiserLabel.text = advertiser
        }
        nativeAdView.advertiserView = advertiserLabel

        // ----- Star Rating -----
        let starLabel = UILabel()
        starLabel.font = UIFont.systemFont(ofSize: 12)
        starLabel.textColor = UIColor(white: 0.46, alpha: 1)
        starLabel.translatesAutoresizingMaskIntoConstraints = false
        if let rating = nativeAd.starRating {
            starLabel.text = "★ \(rating.doubleValue)"
            starLabel.isHidden = false
        } else {
            starLabel.isHidden = true
        }
        nativeAdView.starRatingView = starLabel

        // ----- Media -----
        let mediaView = GADMediaView()
        mediaView.translatesAutoresizingMaskIntoConstraints = false
        mediaView.mediaContent = nativeAd.mediaContent
        nativeAdView.mediaView = mediaView

        // ----- Body -----
        let bodyLabel = UILabel()
        bodyLabel.font = UIFont.systemFont(ofSize: 14)
        bodyLabel.textColor = UIColor(white: 0.26, alpha: 1)
        bodyLabel.numberOfLines = 3
        bodyLabel.translatesAutoresizingMaskIntoConstraints = false
        if let body = nativeAd.body {
            bodyLabel.text = body
        }
        nativeAdView.bodyView = bodyLabel

        // ----- Store -----
        let storeLabel = UILabel()
        storeLabel.font = UIFont.systemFont(ofSize: 12)
        storeLabel.textColor = UIColor(white: 0.46, alpha: 1)
        storeLabel.translatesAutoresizingMaskIntoConstraints = false
        if let store = nativeAd.store {
            storeLabel.text = store
        }
        nativeAdView.storeView = storeLabel

        // ----- Price -----
        let priceLabel = UILabel()
        priceLabel.font = UIFont.boldSystemFont(ofSize: 14)
        priceLabel.textColor = UIColor(red: 0.3, green: 0.69, blue: 0.31, alpha: 1)
        priceLabel.textAlignment = .right
        priceLabel.translatesAutoresizingMaskIntoConstraints = false
        if let price = nativeAd.price {
            priceLabel.text = price
        }
        nativeAdView.priceView = priceLabel

        // ----- CTA -----
        let ctaButton = UIButton(type: .system)
        ctaButton.backgroundColor = UIColor(red: 0.08, green: 0.4, blue: 0.75, alpha: 1)
        ctaButton.setTitleColor(.white, for: .normal)
        ctaButton.titleLabel?.font = UIFont.boldSystemFont(ofSize: 13)
        ctaButton.contentEdgeInsets = UIEdgeInsets(top: 0, left: 16, bottom: 0, right: 16)
        ctaButton.layer.cornerRadius = 6
        ctaButton.translatesAutoresizingMaskIntoConstraints = false
        if let cta = nativeAd.callToAction {
            ctaButton.setTitle(cta, for: .normal)
        }
        nativeAdView.callToActionView = ctaButton

        // ----- Layout -----

        nativeAdView.addSubview(attributionLabel)
        nativeAdView.addSubview(iconView)
        nativeAdView.addSubview(headlineLabel)
        nativeAdView.addSubview(advertiserLabel)
        nativeAdView.addSubview(starLabel)
        nativeAdView.addSubview(mediaView)
        nativeAdView.addSubview(bodyLabel)
        nativeAdView.addSubview(storeLabel)
        nativeAdView.addSubview(priceLabel)
        nativeAdView.addSubview(ctaButton)

        NSLayoutConstraint.activate([
            // Attribution
            attributionLabel.topAnchor.constraint(equalTo: nativeAdView.topAnchor, constant: 12),
            attributionLabel.leadingAnchor.constraint(equalTo: nativeAdView.leadingAnchor, constant: 12),

            // Icon
            iconView.topAnchor.constraint(equalTo: attributionLabel.bottomAnchor, constant: 8),
            iconView.leadingAnchor.constraint(equalTo: nativeAdView.leadingAnchor, constant: 12),
            iconView.widthAnchor.constraint(equalToConstant: 40),
            iconView.heightAnchor.constraint(equalToConstant: 40),

            // Headline
            headlineLabel.topAnchor.constraint(equalTo: iconView.topAnchor),
            headlineLabel.leadingAnchor.constraint(equalTo: iconView.trailingAnchor, constant: 10),
            headlineLabel.trailingAnchor.constraint(equalTo: nativeAdView.trailingAnchor, constant: -12),

            // Advertiser
            advertiserLabel.topAnchor.constraint(equalTo: headlineLabel.bottomAnchor, constant: 2),
            advertiserLabel.leadingAnchor.constraint(equalTo: headlineLabel.leadingAnchor),
            advertiserLabel.trailingAnchor.constraint(equalTo: headlineLabel.trailingAnchor),

            // Star
            starLabel.topAnchor.constraint(equalTo: advertiserLabel.bottomAnchor, constant: 2),
            starLabel.leadingAnchor.constraint(equalTo: headlineLabel.leadingAnchor),

            // Media
            mediaView.topAnchor.constraint(equalTo: iconView.bottomAnchor, constant: 10),
            mediaView.leadingAnchor.constraint(equalTo: nativeAdView.leadingAnchor, constant: 12),
            mediaView.trailingAnchor.constraint(equalTo: nativeAdView.trailingAnchor, constant: -12),
            mediaView.heightAnchor.constraint(equalToConstant: 160),

            // Body
            bodyLabel.topAnchor.constraint(equalTo: mediaView.bottomAnchor, constant: 8),
            bodyLabel.leadingAnchor.constraint(equalTo: nativeAdView.leadingAnchor, constant: 12),
            bodyLabel.trailingAnchor.constraint(equalTo: nativeAdView.trailingAnchor, constant: -12),

            // Store
            storeLabel.topAnchor.constraint(equalTo: bodyLabel.bottomAnchor, constant: 6),
            storeLabel.leadingAnchor.constraint(equalTo: nativeAdView.leadingAnchor, constant: 12),

            // Price
            priceLabel.centerYAnchor.constraint(equalTo: storeLabel.centerYAnchor),
            priceLabel.leadingAnchor.constraint(equalTo: storeLabel.trailingAnchor, constant: 8),
            priceLabel.trailingAnchor.constraint(equalTo: nativeAdView.trailingAnchor, constant: -12),

            // CTA
            ctaButton.topAnchor.constraint(equalTo: storeLabel.bottomAnchor, constant: 8),
            ctaButton.leadingAnchor.constraint(equalTo: nativeAdView.leadingAnchor, constant: 12),
            ctaButton.heightAnchor.constraint(equalToConstant: 38),
            ctaButton.bottomAnchor.constraint(equalTo: nativeAdView.bottomAnchor, constant: -12),
        ])

        nativeAdView.nativeAd = nativeAd
        return nativeAdView
    }
}
